### ISSUE TITLE
Possible loss of log messages when syslog-ng.persist file has cursor "in the future".

### ISSUE THREAD

**Issue by faithlesstomas** [balabit/syslog-ng#1570]
## Issue

**Description**
When syslog-ng checks cursor position from syslog-ng.persist file it does not check if saved cursor is "in the future" with respect to the system (i.e. time in the cursor is in the future with the respect to the system time). This causes logging glitch when syslog-ng is started (syslog-ng waits for the "time to come" to continue logging) when there is no systemd journal entries with the given "future time" or later. This is because sd_journald_seek_cursors() seeks to the next closest entry (in terms of time) if no entry matches.

**Possible fix**
Use sd_journal_test_cursor() to test the validity of the cursor.

**Side note**
This issue won't occur normally unless someone or something mess up with system time and systemd journal. Consider following example, how one can reproduce this bug:
1. stop syslog-ng and systemd-journald services
2. cleanup/remove journal logs
3. set the time in some past with respect to given time
4. reboot machine

## Version of syslog-ng

3.10.0

## Platform

linux debian

**Comment by faithlesstomas** [comment]
I see that proposed bugfix for this issue (#1612) have been rejected/removed from source code.  For me, especially after discussing this issue against systemd-journald API with Lennart Poettering (https://github.com/systemd/systemd/issues/6213), this bug seems to be quite severe. Is there any possibility to revise this again? Or is there any other solution?

**Comment by lbudai** [comment]
Hmm... there is a documentation bug around this `sd_journal_test_cursor` (or, if someone prefers, it's not as precise as it should be...).

The original patch calls `test_cursor` in the wrong place: it requires `sd_journal_next` (as in case of `sd_journal_get_cursor`). I'll explain it, but before that, let's check what the documentation says:

* Doc says:

```
Note that sd_journal_get_cursor() will not work before sd_journal_next(3) (or related call) has
       been called at least once, in order to position the read pointer at a valid entry.
```

* and in another point:

```
sd_journal_test_cursor() may be used to check whether the current position in the journal
       matches the specified cursor. This is useful since cursor strings do not uniquely identify an
       entry: the same entry might be referred to by multiple different cursor strings, and hence
       string comparing cursors is not possible. Use this call to verify after an invocation of
       sd_journal_seek_cursor(3) whether the entry being sought to was actually found in the journal
       or the next closest entry was used instead
```

`sd_journal_seek_cursor` not depends on `sd_journal_next` (or related calls...) and after this call you have to set the position, otherwise `sd_journal_test_cursor` will fail with EADDRNOTAVAIL.

* Why?

Let's check the code:
First, relevant part from `sd_journal_get_cursor`:

```
_public_ int sd_journal_get_cursor(sd_journal *j, char **cursor) {
        Object *o;
        int r;
        char bid[33], sid[33];

        assert_return(j, -EINVAL);
        assert_return(!journal_pid_changed(j), -ECHILD);
        assert_return(cursor, -EINVAL);

        if (!j->current_file || j->current_file->current_offset <= 0)
                return -EADDRNOTAVAIL;

```

then from`sd_journal_test_cursor`:

```
_public_ int sd_journal_test_cursor(sd_journal *j, const char *cursor) {
        int r;
        Object *o;

        assert_return(j, -EINVAL);
        assert_return(!journal_pid_changed(j), -ECHILD);
        assert_return(!isempty(cursor), -EINVAL);

        if (!j->current_file || j->current_file->current_offset <= 0)
                return -EADDRNOTAVAIL;


```

(side note: 'nice' copy-paste...)

The file and the offset is set by the `set_location`:

```
static void set_location(sd_journal *j, JournalFile *f, Object *o) {
        assert(j);
        assert(f);
        assert(o);

        init_location(&j->current_location, LOCATION_DISCRETE, f, o);

        j->current_file = f;
        j->current_field = 0;

        /* Let f know its candidate entry was picked. */
        assert(f->location_type == LOCATION_SEEK);
        f->location_type = LOCATION_DISCRETE;
}

```

And it is called by `sd_journal_next`...

* So the fix will be simple, but would be nice to report back if it works or not.
If our test will pass, we will release it as-is.

**Comment by MrAnno** [comment]
@faithlesstomas #1612 has been included in the new [syslog-ng v3.16.1](https://github.com/balabit/syslog-ng/releases/tag/syslog-ng-3.16.1) release. Thank you for the fix and your patience.

**Comment by faithlesstomas** [comment]
@lbudai @MrAnno Thank you too!
